### DISCUSSION TITLE
Setting inverter limit with PHP via Web API

### DISCUSSION THREAD

**Discussion by Forhorse88** [tbnobody/OpenDTU#1701] [Q&A] [answered]
Hello,

I'm trying to set a non peristant power limit to my inverter using web API and PHP
But all I get is the following answer

>  {"type":"warning","message":"No values found!","code":1002}

there is my code :
```php
	$ip = "192.168.0.24";
	$url = 'http://'.ip.'/api/limit/status';
	$curl = curl_init();
	curl_setopt($curl, CURLOPT_URL, $url);
	curl_setopt($curl, CURLOPT_RETURNTRANSFER, true);
	curl_setopt($curl, CURLOPT_CONNECTTIMEOUT, 15);

	$datas = curl_exec($curl); 
		
	curl_close($curl);
	
	$parsed_json = json_decode($datas, true);

	$x=1;

	foreach ($parsed_json as $key=>$values){
		$onduleurs[$x] = $key;
		foreach ($values as $ikey=>$ivalues){
			$scenario->setLog($x.":".$ikey."=".$ivalues); //for Jeedom logs
		}
		$x++;
	}
//Everthing work well to this point
//$onduleurs[1] contain the serial number of my only inverter (for now)

	$url = 'http://'.$ip.'/api/limit/config';
	$postdata = "data={\"serial\":\"$onduleurs[1]\", \"limit_type\":1, \"limit_value\":50}";
	$username = "admin";
	$password = "openDTU42";
	$ch = curl_init($url); 
	curl_setopt($ch, CURLOPT_POST, 1);
	curl_setopt($ch, CURLOPT_USERPWD, $username . ":" . $password);
	curl_setopt($ch, CURLOPT_POSTFIELDS, $postdata);
	curl_setopt($ch, CURLOPT_RETURNTRANSFER, 1); 
	curl_setopt($ch, CURLOPT_HTTPHEADER, array('Content-Type: application/json'));
	$result = curl_exec($ch);
	curl_close($ch);

``` 
I'm not very good at PHP or programming in general
What i'm doing wrong ?
Thank you

**Comment by MetaChuh** [comment, marked as answer]
@Forhorse88 

if you still need a solution, here's your code, corrected and working:

long version, based on your original code example:

```php
$ip="your_opendtu_ip";
$username = "admin";
$password = "your_password";
$onduleurs[1]="your_inverter_serial";
$limit_value=50;

$url = "http://$ip/api/limit/config";
$postdata = "data={'serial':'$onduleurs[1]', 'limit_type':1, 'limit_value':$limit_value}";

$ch = curl_init();

curl_setopt($ch, CURLOPT_URL,"$url");
curl_setopt($ch, CURLOPT_USERPWD, "$username:$password");
curl_setopt($ch, CURLOPT_POST, true);
curl_setopt($ch, CURLOPT_POSTFIELDS, $postdata);
curl_setopt($ch, CURLOPT_RETURNTRANSFER, true);

$result = curl_exec($ch);

curl_close($ch);

// response output:
echo "<pre>php curl - issue #1701 fix:\n";
echo "result: $result\n";
```

alternative short version, if you are on linux, mac, dsm or similar:
(platform simplification, reducing the code to a short one-liner)

```php
$result=`curl --request POST --data "$postdata" --user "$username:$password" "$url"`;

// response output:
echo "result: $result";
```

greetings

**Comment by Forhorse88** [comment]
It's work !
Thank you !

Do you know if it possible to set the limite of multiple inverters with one request ($postdata with multiple serial ?)

**Reply by MetaChuh** [reply]
no, you have to do a separate post for each inverter

**Reply by Forhorse88** [reply]
OK, I understand. Thanks again for your help

**Comment by markok12345** [comment]
Hi,

I have same problem with Shelly script. I tried this code, but I got same response. Can somebody help me what is wrong?

``` 
function ParseResponse(res, error_code, error_msg, ud) {
    if (error_code != 0) {
        print("Request error: ", error_code, error_msg);
    } else {
        if (res.code === 200) {
            let response = JSON.parse(res.body);
            print(response);
        } else {
            print('Request failed: ', res.code);
        }
    }
}
let username = 'xxxxxxx'
let password = 'xxxxxxx'
let urlToCall = 'http://' + username + ':' + password + '@192.168.1.113/api/limit/config'
let payload = 'data={"serial":"1144822xxxx", "limit_type":1, "limit_value":50}'

Shelly.call("HTTP.POST", { url: urlToCall, body: payload, timeout: 15, ssl_ca: "*" }, ParseResponse); 
```
Edit, I tried paste this as code, but I missing something
Edit 2, MetaChuh, Thank you for your help. Now I can paste code right way.

**Reply by MetaChuh** [reply]
i'm not familiar with shelly script yet, so i don't know.
but the idea to control directly from eg. a 3em without an extra (web) server looks interesting.

to paste your multi line code, you have to paste it between two lines of backticks ` ``` `

example:
` ``` `
```
your code
```
` ``` `

**Comment by markok12345** [comment]
I tried curl command Windows command prompt too and there is something what i can't understand. When I send GET command like " curl http://192.168.10.10/api/livedata/status " everything looks good. 

When I send POST command like " curl -u "username:pwd" http://192.168.10.10/api/limit/config -d 'data={"serial":"1144822xxxxx", "limit_type":1, "limit_value":50} " 

I get response message " {"type":"warning","message":"No values found!","code":1002}curl: (3) URL rejected: Port number was not a decimal number between 0 and 65535 curl: (3) unmatched close brace/bracket in URL position 15: limit_value:50}

Username, password, IP address and inverter serial number are my own when I testing this. What this error message means?

**Reply by MetaChuh** [reply]
your syntax problem is at `-d 'data... "`
if you open a string with a `'` you have to close it with a `'`
you can not mix opening and closing brackets `'`and `"`

**Reply by MetaChuh** [reply]
affirmation: just tested in windows with curl, your brackets were the only issue.

for your convenience, here's the corrected code line, according to your example:

```
curl --user "username:pwd" http://192.168.10.10/api/limit/config --data "data={'serial':'1144822xxxxx', 'limit_type':1, 'limit_value':50}"
```

albeit, i recommend to explicitly state the request type with `--request POST`, for compatibility to future curl versions

recomended:
```
curl --user "username:pwd" http://192.168.10.10/api/limit/config --request POST --data "data={'serial':'1144822xxxxx', 'limit_type':1, 'limit_value':50}"
```

**Comment by MetaChuh** [comment]
@Forhorse88 
if the original topic is solved for you, please mark it as answered.
this helps users to find and differentiate between questions with or without decisive solution.
thx.
